<a href="https://colab.research.google.com/github/Sulaiman29/BDA_HW1/blob/main/BDA2025Spring_HW1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### <div class="alert alert-block alert-info"><b></b>
<h1><center> Homework 01</center></h1>
<h2><center> Data Exploration and Customer Segmentation</center></h2>    
<h2><center> MTAT.03.319 - Business Data Analytics</center></h2>
<h2><center> University of Tartu - Spring 2025</center></h2>
</div>

# Homework instructions

- Please provide the names and student IDs of the team-members (Maximum 2 person) in the field "Team mates" below. If you are not working in a team please insert only your name and student ID.

- Please provide code where ever applicable.

- The accepted submission format is .ipynb file. Please make sure that the privacy settings for the file is public so we can access your code.

- The submission will automatically close on <font color='red'>**16 March at 23:59**</font>, so please make sure to submit before the deadline.

- ONLY one of the teammates should submit the homework and in the submission description the other person's Name and Student ID must be entered. We will grade the homework and the marks and feedback is applied for both the team members. So please communicate with your team member about marks and feedback if you are submit the homework.

- If a question is not clear, please ask us in Moodle ONLY.

- After you have finished solving the Homework, please restart the Kernel and run all the cells to check if there is any persisting issues.

- Plagiarism is <font color='red'>**PROHIBITED**</font>. Any form of plagiarism will be dealt according to the university policy (https://ut.ee/en/content/academic-fraud).

- <font color='red'>**DO NOT CHANGE THE TEMPLATE**</font>

- <font color='red'>**Restart the Kernel and Run all the cells once again after you are done.**</font>
This will ensure that all the cells run without error. You will find an option in the top menu bar under Kernel tab.

**<h2><font color='red'>Team mates:</font></h2>**


<font color='red'>Name: </font>Muhammad Sulaiman Javed &emsp; <font color='red'>Student ID: </font> ABC123456


<font color='red'>Name: </font> &emsp; <font color='red'>Student ID: </font>




In [3]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


# Introduction

* In this homework you are going to use Python; specifically Pandas package which is a powerful and handy data analysis and manipulation tool, built on top of the Python programming language. In addition, there are questions related to clustering as well. The tasks are relevant to the tasks from the practice session. So, please make sure to go through the practise session carefully before you attempt to solve this homework.

### The homework is divided into four sections and the points are distributed as below:
<pre>
- Python and Pandas    -> 1.5 points
- Understanding data   -> 2.5 points
- Data Preparation     -> 1.5 points
- Exploring dataset    -> 2.0 points
- Basic Clustering     -> 6.0 points
_________________________________________
Total                  -> 13.5 points
</pre>

# 1. Python and Pandas (1.5 points)

First, let's import pandas package as pd and read data from the abalone CSV file into a pandas DataFrame.

In [4]:
import pandas as pd

data = pd.read_csv('abalone.csv')

In [5]:
data

Gender  Length  Diameter  Height  Weight  Rings
0        F   0.505     0.385   0.135  0.6185     12
1        F   0.650     0.475   0.165  1.3875      9
2        I   0.520     0.380   0.135  0.5395      8
3        F   0.550     0.425   0.140  0.9520      7
4        M   0.640     0.525   0.180  1.3135     10
..     ...     ...       ...     ...     ...    ...
995      I   0.550     0.415   0.135  0.8095     12
996      I   0.455     0.435   0.110  0.4265      8
997      M   0.660     0.500   0.165  1.3195      9
998      I   0.525     0.400   0.130  0.6455      8
999      M   0.515     0.395   0.135  1.0070      8

[1000 rows x 6 columns]

`Question 1.0` is an example of how you should answer the following questions. First, in the `#TODO` cell you have to write the code representing your solution. And in the answer cell write the answer with your own words. Please make sure your sentences should be clear enough. Try to avoid abbreviations and short sentences.

 **1.0. What is overall memory consumption of the data?** (EXAMPLE)

In [6]:
# TODO - here you have to write the code.
data.memory_usage(deep=True).sum()

98132

* Provide description based on the outcome of the cell above. Please describe as much as possible and as little as required. If you are asked to display a dataframe, you can use the code cells and display the dataframe in the output. You can also use the answer cell (markdown type) for descriptive answers.



**<font color='red'>Answer:</font>** Overall memory consumption: 98132 bytes

 **1.1. How many variables (features) are in the dataset? State the feature names. (0.10 point)** <br>
 Hint: Variables or features are the number of columns in a dataset.

**<font color='red'>Answer:</font>**

**1.2. How many observations (i.e. rows) are in this data frame?** **(0.10 point)**


**<font color='red'>Answer:</font>**

**1.3. Print the first 5 lines from the dataset. What are the values of feature `Rings`?** **(0.10 point)**

**<font color='red'>Answer:</font>**

**1.4. Extract the last 3 rows of the data frame. What are the values of feature `Weight`?** **(0.10 point)**

**<font color='red'>Answer:</font>**

**1.5. What is the value of diameter in the row 575?** **(0.25 point)**

**<font color='red'>Answer:</font>**

**1.6. What are the mean, median and third (75%) quartile value of the height column?** **(0.25 point)** <br>
Hint: Use pandas describe() method for mean and third quartile values.

**<font color='red'>Answer:</font>**

**1.7. Show the subset of rows from the data frame where `Gender` is F and `Weight` values are above 0.75. What is the mean of `Diameter` in this subset?** **(0.30 point)**

**<font color='red'>Answer:</font>**

**1.8. Show the subset of rows from the dataframe where `Rings` values are exactly 18? <br>
Find the minimum value of `Length` of this subset** **(0.30 point)**

**<font color='red'>Answer:</font>**

# 2.  Understanding Data (2.5 point)

Data loading, describing, exploring, and verifying the quality of the data are <b>pivotal</b> to any data analysis task. <br>
We will work with a new dataset, 'adult.csv',in this section. You will be asked to perform the mentioned tasks on this dataset.<br>

**2.0. Load data from the  `adult.csv` file.** <br>
Hint 1: The extension of the filename may not always reflect the structure of data precisely. Are the values comma-separated (csv) or not?

In [ ]:
import numpy as np
import pandas as pd


data = pd.read_csv('adult.csv', sep='\t', index_col=0)
data.head()

age          workclass   education          occupation  capital.gain  \
X                                                                        
0  39          State-gov   Bachelors        Adm-clerical          2174   
1  50   Self-emp-not-inc   Bachelors     Exec-managerial             0   
2  38             privat     HS-grad   Handlers-cleaners             0   
3  53             privat        11th   Handlers-cleaners             0   
4  28             privat   Bachelors      Prof-specialty             0   

   capital.loss  native.country  salaries jobsatisfaction  male  female  
X                                                                        
0             0   United-States   43136.0               0   1.0     NaN  
1             0    UnitedStates   46209.0               7   1.0     NaN  
2             0   United-States   28937.0               6   1.0     NaN  
3             0   United-States   33658.0               6   1.0     NaN  
4             0            Cuba   34372.0              12   NaN     1.0

**2.1. Name all the columns in this dataset and describe the meaning of each column in one sentence. (0.25 points)** <br>

**<font color='red'>Answer:</font>**

####  2.1.1.Visualization is one of the easiest ways to get a good overview of the data, notice any peculiarity, or outliers. <br>In the following subtasks, you will have to visualize the distribution of the values for different features using appropriate diagrams.

In [ ]:
# Matplotlib is a Python 2D plotting library which produces publication quality figures in a variety of hardcopy formats
# and interactive environments across platforms. Matplotlib can be used in Python scripts, the Python and IPython shells,
# Jupyter notebook, web application servers, and for graphical user interface toolkits.
import matplotlib.pyplot as plt

# allows to output plots in the notebook
%matplotlib inline

# Set the default style
plt.style.use('ggplot')

As an example, here is the frequency histogram of "age" column, where age is an attribute of type **numeric**:

In [ ]:
# Columns may contain non-numeric values, errors or missing values. Therefore, non-numeric values must be dealt with.
if data is not ...:
    pd.to_numeric(data['age'], errors='coerce').hist(bins=30) # ‘coerce’ -> invalid parsing will be set as NaN
else:
    print('Please define `data` in earlier subtasks')

**2.2. Specify the type of `workclass` attribute (nominal / ordinal / numeric). Choose an appropriate chart to visualise the frequency distribution for the values of `workclass`. (0.25 points)** <br>
Hint: Do not forget the data preprocessing part. First look for any unusual value in the specific column and deal with it appropriately. You do not necessarily need to eyeball each of the rows. But you can think of other ways perhaps. The original dataset uses '?' for missing values

**<font color='red'>Attribute type :</font>**

**2.3. Specify the type of `education` attribute (nominal / ordinal / numeric). Choose an appropriate chart to visualise the frequency distribution for the values of `education`. (0.25 points)** <br>

**<font color='red'>Attribute type :</font>**

**2.4. Specify the type of `occupation` attribute (nominal / ordinal / numeric). Choose an appropriate chart to visualise the frequency distribution for the values of `occupation`. (0.25 points)**

**<font color='red'>Attribute type :</font>**

**2.5. Specify the type of `native.country` attribute (nominal / ordinal / numeric). Choose an appropriate chart to visualise the frequency distribution for the values of `native.country`. (0.25 points)**

**<font color='red'>Attribute type :</font>**

**2.6. Specify the type of `salaries` attribute (nominal / ordinal / numeric). Choose an appropriate chart to visualise the frequency distribution for the values of `salaries`. (0.25 points)**

**<font color='red'>Attribute type :</font>**

**2.7. Attributes `male` and `female`. Specify the types of these attributes (nominal / ordinal / numeric). Choose an appropriate chart type to visualise the relative frequency of positive values in attribute `male` and positive values in attribute `female` (relative frequency of males and females). (0.25 points)**

**<font color='red'>Attribute type :</font>**

**2.8. Count the number of missing values in each attribute in the dataset and report the percentage of missing values of each attribute (the original file uses '?' for missing values). (0.25 points)**

**<font color='red'>Answer:</font>**

<strong>2.9. Please provide a descriptive statistics for the people who work(`workclass`) in private sector. <br>Separately answer the following questions, <br> What are their mean and median salaries (`salaries`), job satisfaction (`jobsatisfaction`)? <br> What are their highest and lowest level of education (`education`)? <br> Provide the percentage of `male` and `female` employee distribution of these people.  (0.5 points)<strong>

**<font color='red'>Answer:</font>**

# 3.  Data preparation (1.5 points)

Preparing the data for analysis is the most crucial part of data analysis. In this exercise, we will clean up the dataset 'adult.csv'. The dataset was purposefully made untidy so that you can practice cleaning it up.<br>

First, let's make a deep copy of the original DataFrame. Later we will compare it to the cleaned/modified DataFrame.

In [ ]:
if data is not ...:
    original_data = data.copy(deep=True) # Make a deep copy, including a copy of the data and the indices

**3.1. Sometimes values include leading and/or trailing spaces. Remove leading and trailing whitespace from strings. Count the number of rows that were changed in the dataset. (0.30 points)** <br>

In [ ]:
#Perform manipulations in the 'data' dataframe. So that later you can use the code block below to compare


Use code below to count how many rows were affected.

In [ ]:
# Count differences
if data is not ...:
    data_all = pd.concat([original_data, data]).drop_duplicates()
    diff = data_all.shape[0] - data.shape[0]
    print ('Difference: ' + str(diff))

**<font color='red'>Answer:</font>**

**3.2. Replace all non-sensical values in `data` with np.nan from numpy package to denote missing values (Original file uses '?' for missing values). Fix the typos in the dataset. Justify all the changes that you do, explaining why you are sure that this is a correct change and what you think the cause for such error. (0.60 points)**  <br>

**<font color='red'>Answer:</font>**

**3.3. Count the rows which were changed in the subtask 3.2 (0.30 points)** <br>

In [ ]:
data_all = #TODO
diff = data_all.shape[0] - data.shape[0]
print ('Difference: ' + str(diff))

**<font color='red'>Answer:</font>**

**3.4. Report values that look suspicious. By suspicious we mean that you are not sure if those values are wrong or right. This is an open-ended question, use your rational judgement. (0.30 points)** <br>

**<font color='red'>Answer:</font>**

# 4.  Exploring dataset (2 points)

Gathering interesting facts about the `adult.csv` through visualization without complex data analysis.

**4.1. Convert the attribute `Education` into type `category` and introduce the order for this attribute based on the education level (e.g. Masters is higher education than Bachelors; use your best guess in ordering those levels for which the order is not obvious). Create a bar chart visualising the relative frequencies of values. (0.5 Points)**

**<font color='red'>Answer:</font>**

**4.2.  Create and print a Pandas data frame where each row stands for an occupation, each column stands for a level of education, and the cells in the table contain the average salary of people with the corresponding occupation and education level.(0.5 Points)** <br>
Hint 1: You can use crosstab, see also aggfunc from the pandas documentation.

**<font color='red'>Answer:</font>**

**4.3. Make a separate plot for each row (occupation) of the table created in subtask 4.2. Choose the type of the plot to convey the information as good as you can.(0.5 Points)** <br>

**<font color='red'>Answer:</font>**

**4.4. List 3 interesting facts that you can read out of these plots from subtask 4.3. (0.5 points)** <br>

**<font color='red'>Fact 1:</font>**

**<font color='red'>Fact 2:</font>**

**<font color='red'>Fact 3:</font>**

# 5. Basic Clustering (6 points)
We will learn RFM (Recency, Frequency, Monetary value) technique and basic clustering algorithms: K-means, DBSCAN, and Hierarchical clustering in this section.



## 5.1 RFM (1.5 points)
We will apply the RFM technique in this section similar to the practice session. The dataset for this task belongs to http://donorschoose.org/.

The web platform http://donorschoose.org helps projects in collecting donations. The donators have the option to give 15% of what they donate to this webplatform. The dataset contains the following fields:

- Donor ID - Unique identifier of donors
- Donation Included Optional Donation - The donor has agreed(Yes/No) to give 15% of donation amount to Donoschoose.org
- Donation Amount - Total amount donated
- Donor Cart Sequence - The ranking of the donaton project in the donors list
- Donation Received Date - Date of online payment


**5.1.1**
**Please load the dataset Donations.csv. <br>The field ```Donation Included Optional Donation``` shows if the donor agreed (Yes/No) to give 15% of donation amount to Donoschoose.org. <br>Update the field "Donation Amount" so that it represents the correct amount of donation; which means if `Donation Included Optional Donation`=yes, calculate the `Donation Amount`. (0.25 points)**

**<font color='red'>Answer:</font>**

**5.1.2 For understanding `Recency`, consider 2018-05-31 23:00:00 as the reference date
(we assume this is report date).<br> Calculate the number of days between this date and the date of last purchase  for each customer. How long the donor with ID equal to e0fe4d9b8def8a71635e65ba4ff5ef40 has been inactive ? (0.25 points)**

**<font color='red'>Answer:</font>**

**5.1.3 For understanding `Frequency`, calculate how frequently (how many times) a user has made donations.<br> Assume that we want to recognize the person with the highest frequency by displaying her/ his name on the webpage, please find the donor ID for the highest frequency available. (0.25 points)**

**<font color='red'>Answer:</font>**

**5.1.4 For understanding `Monetary Value`, calculate the amount of donation. <br> What is the highest donation amount and what is the donor ID of that donation? <br> Is donor ID from from this the same with donor ID from the highest frequency? (0.25 points)**

**<font color='red'>Answer:</font>**

**5.1.5 Create a column `RFM_Score` with the quantiles of Recency, Frequency, Monetary value. Interpret the value of `RFM_Score`. (0.50 points)**

**<font color='red'>Answer:</font>**

## 5.2 K-means Clustering (1.5 points)
We will learn to apply K-means clustering algorithm in this section.

**5.2.1 Plese load the `customers.csv` dataset and provide a descriptives of the dataset. (0.05 points)** <br>
*Make sure to pass `encoding = "ISO-8859-1"` in the pandas read_csv() function*

**<font color='red'>Answer:</font>**

**5.2.2 Visualize the relation of `Age` and `Annual Income (k$)` of customers in the dataset. You can use your perefered plotting method to show the result. Could you explain the result if there is any relation among the variables? (0.10 points)**

**<font color='red'>Answer:</font>**

**5.2.3 Visualize the relation of `Age` and `Spending Score (1-100)` of customers in the dataset. You can use your perefered plotting method to show the result. Could you explain the result if there is anything interesting going on? (0.10 points)**

**<font color='red'>Answer:</font>**

**5.2.4 Visualize the relation of `Annual Income (k$)` and `Spending Score (1-100)` of customers in the dataset. (0.10 points)**

**<font color='red'>Answer:</font>**

If you successfully plotted the diagram of 5.2.4, you should be able to see how the instances(points) are dispersed; clearly there are some groups where the instances are close to each other. Unfortunately computers cannot just look at the plot, guess on the clusters and their boundaries. <br>In order to make machine understand the clusters, we will apply K-means, a very popular and easy algorithm. <br><br>
But before applying K-means we have to know on which columns we will apply the K-means algorithm. For that, we have to perform data preprocessing. Since the above plot shows a relation between `Anunual Income (k$)` and `Spending Score (1-100)`, it is wise just to pick those two and proceed. <br><br>

However, there are drawbacks of K-means algorithm. It does not inherently know how many possible clusters are available in the dataset. So, we have to set the parameter manually before we run the algorithm. One of the ways to know number clusters is to apply elbow method. The elbow method has been discussed in the lecture. Now it is your turn to apply elbow method on our dataset.

**5.2.5 Select `Anunual Income (k$)` and `Spending Score (1-100)` to create a new dataset for the purpose of elbow method and K-means. Apply elbow method on the new dataset and plot the result. (0.5 points)**

In [ ]:
CRED = '\033[91m'
CEND = '\033[0m'

#packages you need to do the tasks
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
from sklearn.cluster import KMeans

#funtion to plot elbow method
def plot_elbow(x,y):
    if len(y)!=0:

        plt.figure(figsize=(16,8))
        plt.plot(K, dist, 'bx-')
        plt.xlabel('Number of K')
        plt.ylabel('Distortion')
        plt.title('Elbow Method')
        plt.show()
    else:
        raise NotImplementedError("Oops! There is some error, implement the elbow method correctly!")


#data preprocessing



##### YOUR DATA PREPROCESSING CODE STARTS ##### (please do not delete this line)

##### YOUR DATA PREPROCESSING CODE ENDS ##### (please do not delete this line)





dist = []            #Store the distortion in dist variable.
K = range(1,10)      #we can check upto 10 clusters, you can change the value if you want to experiment.




##### YOUR ELBOW METHOD CODE STARTS ##### (please do not delete this line)

##### YOUR ELBOW METHOD CODE ENDS ##### (please do not delete this line)




#plot of elbow method
try:
    plot_elbow(K, dist)
except NotImplementedError as e:
    print(CRED + "TODO: " + str(e) + CEND)

TODO: Oops! There is some error, implement the elbow method correctly!


**<font color='red'>Answer:</font>**

**5.2.6 If you could successfully apply elbow method, pick the optimal value of K and pass the value to the k-means algoithm you are about to apply.
<br> Please apply the K-means algorithm with `max_iter=1`. Pass the argument to the K-means function.
<br>Do not forget to visualize the clustering result. <br>Report the cluster centroids(0.5 points)**

In [ ]:
##### YOUR K-MEANS CODE STARTS ##### (please do not delete this line)

##### YOUR K-MEANS CODE ENDS ##### (please do not delete this line)

**<font color='red'>Answer:</font>**

**5.2.7 Please apply the K-means algorithm again with `max_iter=10`.
<br>Do not forget to visualize the clustering result. <br>Report the cluster centroids(0.05 points)**

**<font color='red'>Answer:</font>**

**5.2.8 Is there a difference between the centroids with different max_iter value? Justify your answer. (0.1 points)**

**<font color='red'>Answer:</font>**

## 5.3 DBSCAN (1.5 points)
We will apply DBSCAN on the same dataset used for K-means in this section.

**5.3.1 Apply DBSCAN on the same dataset you used to train K-means. Play with the parameter, especially `eps`, and get the best clustering result based on the visualitation which means you have to train and visualize multiple times with a range of values of `eps`. You do not have to check an infinite list of values, pick 4 values and check the visualization after clustering with them. Report the one you find to be the best. (1 points)**

**<font color='red'>Answer:</font>**

**5.3.2 Is the result better than K-means? Justify your answer. Make sure you clearly address the reason why it performed better or worse than K-means. Please answer the advantage and disadvantage of both DBSCAN and K-means. (0.5 points)**

**<font color='red'>Answer:</font>**

## 5.4 Hierarchical clustering (1.5 points)
We will apply Hierarchical clustering in this section on the same dataset of K-means. The reason we are repeating the task with different clustering algorithm is to ensure that you understand the nuances of different clustering algorithms and know when to use what.

**5.4.1 Like K-means, we have to know the number of clusters required for performing hierarchical clustering. One such way is through plotting dendograms. Please plot a dendogram on the same dataset you used for k-means and DBSCAN. (0.5 points)**

**<font color='red'>Answer:</font>**

**5.4.2 Explain how you got the optimum number of clusters in the answer section. (0.25 points)**

<font color='red'>**Answer:**</font>

**5.4.3 Apply hierarchical clustering on the dataset using the optimum number of clusters you got from previous task and visualize the result. (0.5 points)**

**<font color='red'>Answer:</font>**

**5.4.4 Based on all three clustering, take the best result and explain.
In the explanation address how you would describe the customer behavior, are they miser or spendthrift or something else? (0.25 points)**

<font color='red'>**Answer:**</font>

## How long (in hours) did it take you to solve the homework?

* Please answer as precisely as you can. It does not affect your points or grade. It is okay, if it took 0.5 hours or 24 hours or more. The collected information will be used to improve future homeworks.

<font color='red'> **Answer:**</font>


## How difficult is the homework?
you can put only number between $0:10$ ($0:$ easy, $10:$ difficult)

<font color='red'> **Answer:**</font>

## Other comments you want to share?

<font color='red'> **Answer:**</font>